### PR TITLE
Metadata Changes

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Introduce output metadata pointer to process the input and store in output pointer for the meta nodes
- Modify clone to either copy or create empty instance
- Remove Resize meta node (As it does not do any processing as rocAL operates on normalized coordinates)
- Fix Box encoder

### CODE REVIEW

File: rocAL/rocAL/include/meta_data/bounding_box_graph.h [hunk 1]
```diff
@@ -25,13 +25,15 @@ THE SOFTWARE.
 #include "meta_data_graph.h"
 #include "meta_node.h"
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/meta_data/bounding_box_graph.h:27]: Add a comment here stating the use of union.

File: rocAL/rocAL/include/meta_data/bounding_box_graph.h [hunk 2]
```diff
@@ -25,13 +25,15 @@ THE SOFTWARE.
...
 
 class BoundingBoxGraph : public MetaDataGraph
 {
 public:
-    void process(pMetaDataBatch meta_data) override;
-    void update_random_bbox_meta_data(pMetaDataBatch meta_data, decoded_image_info decoded_image_info,crop_image_info crop_image_info) override;
-    void update_box_encoder_meta_data(std::vector<float> *anchors, pMetaDataBatch full_batch_meta_data ,float criteria, bool offset , float scale, std::vector<float>& means, std::vector<float>& stds) override;
+    void process(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data) override;
+    void update_random_bbox_meta_data(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data, decoded_image_info decoded_image_info,crop_image_info crop_image_info) override;
+    void update_box_encoder_meta_data(std::vector<float> *anchors, pMetaDataBatch full_batch_meta_data ,float criteria, bool offset , float scale, std::vector<float>& means, std::vector<float>& stds, float *encoded_boxes_data, int *encoded_labels_data) override;
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/meta_data/bounding_box_graph.h:37]: Do we need to pass pointers? Can we get reference?

File: rocAL/rocAL/source/meta_data/meta_node_crop_resize.cpp [hunk 3]
```diff
@@ -29,7 +29,7 @@ void CropResizeMetaNode::initialize()
     _y2_val.resize(_batch_size);
 }
 
-void CropResizeMetaNode::update_parameters(MetaDataBatch* input_meta_data)
+void CropResizeMetaNode::update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data)
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/meta_data/meta_node_crop_resize.cpp:32]: mem copy is not removed in this function

File: rocAL/rocAL/source/meta_data/meta_node_resize_crop_mirror.cpp [hunk 4]
```diff
@@ -30,7 +30,7 @@ void ResizeCropMirrorMetaNode::initialize()
     _mirror_val.resize(_batch_size);
 }
 
-void ResizeCropMirrorMetaNode::update_parameters(MetaDataBatch* input_meta_data)
+void ResizeCropMirrorMetaNode::update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data)
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/meta_data/meta_node_resize_crop_mirror.cpp:33]: memcopy is not removed in this function

File: rocAL/rocAL/include/meta_data/bounding_box_graph.h (hunk 2 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/meta_data/bounding_box_graph.h:37]: Arrays can be passed only as pointers right?

File: rocAL/rocAL/include/meta_data/bounding_box_graph.h (hunk 1 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/meta_data/bounding_box_graph.h:27]: Done

File: rocAL/rocAL/source/meta_data/meta_node_resize_crop_mirror.cpp (hunk 4 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/meta_data/meta_node_resize_crop_mirror.cpp:33]: Done

File: rocAL/rocAL/source/meta_data/meta_node_crop_resize.cpp (hunk 3 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/meta_data/meta_node_crop_resize.cpp:32]: Done

**Comment by fiona-gladwin** [comment]
@shobana-mcw I have resolved all comments

File: rocAL/rocAL/include/meta_data/meta_node_crop.h [hunk 5]
```diff
@@ -32,7 +32,7 @@ class CropMetaNode:public MetaNode
 {
     public:
         CropMetaNode() {};
-        void update_parameters(pMetaDataBatch input_meta_data)override;
+        void update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data)override;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_node_crop.h:35]: Add space before override

File: rocAL/rocAL/include/meta_data/meta_node_crop_resize.h [hunk 6]
```diff
@@ -32,7 +32,7 @@ class CropResizeMetaNode:public MetaNode
 {
     public:
         CropResizeMetaNode() {};
-        void update_parameters(pMetaDataBatch input_meta_data)override;
+        void update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data)override;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_node_crop_resize.h:35]: Add space before override

File: rocAL/rocAL/include/meta_data/meta_node_flip.h [hunk 7]
```diff
@@ -34,7 +34,7 @@ class FlipMetaNode:public MetaNode
 {
     public:
         FlipMetaNode() {};
-        void update_parameters(MetaDataBatch* input_meta_data)override;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_node_flip.h:37]: Add space before override

File: rocAL/rocAL/include/meta_data/meta_node_resize_crop_mirror.h [hunk 8]
```diff
@@ -32,7 +32,7 @@ class ResizeCropMirrorMetaNode:public MetaNode
 {
     public:
         ResizeCropMirrorMetaNode() {};
-        void update_parameters(MetaDataBatch* input_meta_data)override;
+        void update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data)override;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_node_resize_crop_mirror.h:35]: Add space before override

File: rocAL/rocAL/include/meta_data/meta_node_rotate.h [hunk 9]
```diff
@@ -36,7 +36,7 @@ class RotateMetaNode:public MetaNode
 {
     public:
         RotateMetaNode() {};
-        void update_parameters(MetaDataBatch* input_meta_data)override;
+        void update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data)override;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_node_rotate.h:39]: Add space before override

File: rocAL/rocAL/include/meta_data/meta_node_ssd_random_crop.h [hunk 10]
```diff
@@ -33,7 +33,7 @@ class SSDRandomCropMetaNode : public MetaNode
 {
 public:
     SSDRandomCropMetaNode(){};
-    void update_parameters(pMetaDataBatch input_meta_data) override;
+    void update_parameters(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data) override;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_node_ssd_random_crop.h:36]: Add space before override

File: rocAL/rocAL/source/meta_data/meta_node_crop.cpp [hunk 11]
```diff
@@ -48,12 +48,9 @@ void CropMetaNode::update_parameters(pMetaDataBatch input_meta_data)
...
     {
-        auto bb_count = input_meta_data->get_bb_labels_batch()[i].size();
-        int labels_buf[bb_count];
-        BoundingBoxCords box_coords_buf;
-        box_coords_buf.resize(bb_count);
-        memcpy(labels_buf, input_meta_data->get_bb_labels_batch()[i].data(),  sizeof(int)*bb_count);
-        memcpy((void *)box_coords_buf.data(), input_meta_data->get_bb_cords_batch()[i].data(), input_meta_data->get_bb_cords_batch()[i].size() * sizeof(BoundingBoxCord));
+        auto bb_count = input_meta_data->get_labels_batch()[i].size();
+        Labels labels_buf = input_meta_data->get_labels_batch()[i];
+        BoundingBoxCords box_coords_buf = input_meta_data->get_bb_cords_batch()[i];
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/meta_data/meta_node_crop.cpp:53]: Will the below lines work without any issues for double bboxes?

File: rocAL/rocAL/source/meta_data/meta_node_crop_mirror_normalize.cpp [hunk 12]
```diff
@@ -56,12 +56,9 @@ void CropMirrorNormalizeMetaNode::update_parameters(pMetaDataBatch input_meta_da
...
-        BoundingBoxCords coords_buf;
-        coords_buf.resize(bb_count);
-        memcpy(labels_buf, input_meta_data->get_bb_labels_batch()[i].data(),  sizeof(int)*bb_count);
-        memcpy((void *)coords_buf.data(), input_meta_data->get_bb_cords_batch()[i].data(), input_meta_data->get_bb_cords_batch()[i].size() * sizeof(BoundingBoxCord));
+        auto bb_count = input_meta_data->get_labels_batch()[i].size();
+        Labels labels_buf = input_meta_data->get_labels_batch()[i];
+        BoundingBoxCords coords_buf = input_meta_data->get_bb_cords_batch()[i];
         BoundingBoxCords bb_coords;
         BoundingBoxCord temp_box = {0, 0, 1, 1};
         Labels bb_labels;
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/meta_data/meta_node_crop_mirror_normalize.cpp:64]: Will the below lines work without any issues for double bboxes?

File: rocAL/rocAL/source/meta_data/meta_node_crop_resize.cpp [hunk 13]
```diff
@@ -66,14 +64,9 @@ void CropResizeMetaNode::update_parameters(MetaDataBatch* input_meta_data)
...
+        for(uint j = 0; j < bb_count; j++)
         {
-            BoundingBoxCord box;
-            box.l = coords_buf[m++];
-            box.t = coords_buf[m++];
-            box.r = coords_buf[m++];
-            box.b = coords_buf[m++];
-            
+            BoundingBoxCord box = coords_buf[j];
             if (BBoxIntersectionOverUnion(box, crop_box) >= _iou_threshold)
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/meta_data/meta_node_crop_resize.cpp:70]: Will the below lines work without any issues for double bboxes?

File: rocAL/rocAL/source/meta_data/meta_node_flip.cpp [hunk 14]
```diff
@@ -43,12 +43,8 @@ void FlipMetaNode::update_parameters(pMetaDataBatch input_meta_data)
...
     {
         auto bb_count = input_meta_data->get_labels_batch()[i].size();
-        BoundingBoxCords coords_buf;
-        Labels labels_buf;
-        coords_buf.resize(bb_count);
-        labels_buf.resize(bb_count);
-        memcpy(labels_buf.data(), input_meta_data->get_labels_batch()[i].data(),  sizeof(int)*bb_count);
-        memcpy((void *)coords_buf.data(), input_meta_data->get_bb_cords_batch()[i].data(), input_meta_data->get_bb_cords_batch()[i].size() * sizeof(BoundingBoxCord));
+        BoundingBoxCords coords_buf = input_meta_data->get_bb_cords_batch()[i];
+        Labels labels_buf = input_meta_data->get_labels_batch()[i];
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/meta_data/meta_node_flip.cpp:47]: Will the below lines work without any issues for double bboxes?

File: rocAL/rocAL/source/meta_data/meta_node_resize_crop_mirror.cpp [hunk 15]
```diff
@@ -52,15 +52,12 @@ void ResizeCropMirrorMetaNode::update_parameters(MetaDataBatch* input_meta_data)
...
+        auto bb_count = input_meta_data->get_labels_batch()[i].size();
+        Labels labels_buf = input_meta_data->get_labels_batch()[i];
+        BoundingBoxCords box_coords_buf = input_meta_data->get_bb_cords_batch()[i];
         BoundingBoxCords bb_coords;
         BoundingBoxCord temp_box;
-        BoundingBoxLabels bb_labels;
+        Labels bb_labels;
         BoundingBoxCord crop_box;
         _crop_w = _x2_val[i] - _x1_val[i];
         _crop_h = _y2_val[i] - _y1_val[i];
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/meta_data/meta_node_resize_crop_mirror.cpp:63]: Will the below lines work without any issues for double bboxes?

File: rocAL/rocAL/source/meta_data/meta_node_ssd_random_crop.cpp [hunk 16]
```diff
@@ -50,12 +50,9 @@ void SSDRandomCropMetaNode::update_parameters(pMetaDataBatch input_meta_data)
...
-        BoundingBoxCords box_coords_buf;
-        box_coords_buf.resize(bb_count);
-        memcpy(labels_buf, input_meta_data->get_bb_labels_batch()[i].data(),  sizeof(int)*bb_count);
-        memcpy((void *)box_coords_buf.data(), input_meta_data->get_bb_cords_batch()[i].data(), input_meta_data->get_bb_cords_batch()[i].size() * sizeof(BoundingBoxCord));
+        auto bb_count = input_meta_data->get_labels_batch()[i].size();
+        Labels labels_buf = input_meta_data->get_labels_batch()[i];
+        BoundingBoxCords box_coords_buf = input_meta_data->get_bb_cords_batch()[i];
         BoundingBoxCords bb_coords;
         Labels bb_labels;
         BoundingBoxCord crop_box;
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/meta_data/meta_node_ssd_random_crop.cpp:58]: Will the below lines work without any issues for double bboxes?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 17]
```diff
@@ -528,39 +527,41 @@ void MasterGraph::output_routine()
             }
 
             update_node_parameters();
+            pMetaDataBatch output_meta_data = nullptr;
             if(_augmented_meta_data)
             {
+                output_meta_data = _augmented_meta_data->clone(!_augmentation_metanode); // copy the data if metadata is not proessed by the nodes, else create an empty instance
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:533]: Change it to processed

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 18]
```diff
@@ -242,7 +242,7 @@ class MetaDataBatch
         *this += *other;
         return this;
     }
-    virtual std::shared_ptr<MetaDataBatch> clone() = 0;
+    virtual std::shared_ptr<MetaDataBatch> clone(bool copy_contents = true) = 0;
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/meta_data/meta_data.h:245]: I see from master_graph.cpp that we are creating an empty instance if there's no contents in this. But is there an use case for the else condition. If not, can you make the if condition the default?

Comment (fiona-gladwin) [reply rocAL/rocAL/include/meta_data/meta_data.h:245]: There is use for else condition

File: rocAL/rocAL/source/meta_data/bounding_box_graph.cpp [hunk 19]
```diff
@@ -162,66 +156,61 @@ void BoundingBoxGraph::update_box_encoder_meta_data(std::vector<float> *anchors,
...
-                    encoded_labels[anchor_idx] = bb_labels.at(best_idx);
+                    encoded_labels[anchor_idx] = bb_labels[best_idx];
                 }
             }
             else // Not a match
             {
                 if (offset)
                 {
-                    encoded_bb[anchor_idx] = {0, 0, 0, 0};
+                    encoded_bb[anchor_idx].xcycwh = {0, 0, 0, 0};
```
Comment (fiona-gladwin) [inline rocAL/rocAL/source/meta_data/bounding_box_graph.cpp:200]: Can we just have it as `encoded_bb[anchor_idx]= {0, 0, 0, 0};` Will that work here?

File: rocAL/rocAL/source/meta_data/meta_node_ssd_random_crop.cpp (hunk 16 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/meta_data/meta_node_ssd_random_crop.cpp:58]: Will check and issue this in a different PR

**Comment by fiona-gladwin** [comment]
External PR merged
Need to resolve one comment in this
@fiona-gladwin

**Comment by shobana-mcw** [comment]
> External PR merged Need to resolve one comment in this @fiona-gladwin

Which comment needs to be resolved?

**Pull request closed without merging**